docs/JAX_machine_translation.md

```diff
@@ -0,0 +1,502 @@
+---
+jupytext:
+  formats: ipynb,md:myst
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# NLP: JAX Machine Translation
+
+```{code-cell} ipython3
+## adapted from https://keras.io/examples/nlp/neural_machine_translation_with_transformer/, which is itself an adaptation from https://www.manning.com/books/deep-learning-with-python-second-edition
+```
+
+```{code-cell} ipython3
+import os
+
+import pathlib
+import random
+import string
+import re
+import numpy as np
+
+import jax
+import jax.numpy as jnp
+import optax
+
+from flax import nnx
+
+import tiktoken
+```
+
+## Pull down data to temp and extract into memory
+
+There are lots of ways to get this done, but for simplicity and clear visibility into what's happening this is downloaded to a temporary directory, extracted there, and read into a python object with processing.
+
+```{code-cell} ipython3
+import requests
+import zipfile
+import tempfile
+
+url = "http://storage.googleapis.com/download.tensorflow.org/data/spa-eng.zip"
+
+with tempfile.TemporaryDirectory() as temp_dir:
+    temp_path = pathlib.Path(temp_dir)
+    zip_file_path = temp_path / "spa-eng.zip"
+
+    response = requests.get(url)
+    zip_file_path.write_bytes(response.content)
+
+    with zipfile.ZipFile(zip_file_path, "r") as zip_ref:
+        zip_ref.extractall(temp_path)
+
+    text_file = temp_path / "spa-eng" / "spa.txt"
+
+    with open(text_file) as f:
+        lines = f.read().split("\n")[:-1]
+    text_pairs = []
+    for line in lines:
+        eng, spa = line.split("\t")
+        spa = "[start] " + spa + " [end]"
+        text_pairs.append((eng, spa))
+```
+
+## Build train/validate/test pair sets
+We'll stay close to the original tutorial so it's clear how to follow what's the same vs what's different; one early difference is the choice to go with an off-the-shelf encoder/tokenizer in tiktoken. Specifically "cl100k_base" - it has a wide range of language understanding and it's fast.
+
+```{code-cell} ipython3
+random.shuffle(text_pairs)
+num_val_samples = int(0.15 * len(text_pairs))
+num_train_samples = len(text_pairs) - 2 * num_val_samples
+train_pairs = text_pairs[:num_train_samples]
+val_pairs = text_pairs[num_train_samples : num_train_samples + num_val_samples]
+test_pairs = text_pairs[num_train_samples + num_val_samples :]
+
+print(f"{len(text_pairs)} total pairs")
+print(f"{len(train_pairs)} training pairs")
+print(f"{len(val_pairs)} validation pairs")
+print(f"{len(test_pairs)} test pairs")
+```
+
+```{code-cell} ipython3
+tokenizer = tiktoken.get_encoding("cl100k_base")
+```
+
+We strip out puctuation to keep things simple and in line with the original tutorial - the `[` `]` are kept in so that our `[start]` and `[end]` formatting is preserved.
```

Comment:
```suggestion
We strip out punctuation to keep things simple and in line with the original tutorial - the `[` `]` are kept in so that our `[start]` and `[end]` formatting is preserved.
```